File: app/src/main/java/org/ole/planet/myplanet/ui/myhealth/AddExaminationActivity.java
package org.ole.planet.myplanet.ui.myhealth;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.MenuItem;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.EditText;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import com.google.android.flexbox.FlexboxLayout;
import com.google.gson.Gson;
import com.google.gson.JsonObject;

import org.ole.planet.myplanet.R;
import org.ole.planet.myplanet.databinding.ActivityAddExaminationBinding;
import org.ole.planet.myplanet.databinding.ActivityFeedbackDetailBinding;
import org.ole.planet.myplanet.datamanager.DatabaseService;
import org.ole.planet.myplanet.model.RealmMyHealth;
import org.ole.planet.myplanet.model.RealmMyHealthPojo;
import org.ole.planet.myplanet.model.RealmUserModel;
import org.ole.planet.myplanet.service.UserProfileDbHandler;
import org.ole.planet.myplanet.utilities.AndroidDecrypter;
import org.ole.planet.myplanet.utilities.Constants;
import org.ole.planet.myplanet.utilities.DimenUtils;
import org.ole.planet.myplanet.utilities.JsonUtils;
import org.ole.planet.myplanet.utilities.TimeUtils;
import org.ole.planet.myplanet.utilities.Utilities;

import java.util.Arrays;
import java.util.Date;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import fisk.chipcloud.ChipCloud;
import fisk.chipcloud.ChipCloudConfig;
import io.realm.Realm;

public class AddExaminationActivity extends AppCompatActivity implements CompoundButton.OnCheckedChangeListener {
    private ActivityAddExaminationBinding activityAddExaminationBinding;
    Realm mRealm;
    String userId;
    RealmUserModel user;
    RealmUserModel currentUser;
    RealmMyHealthPojo pojo;
    RealmMyHealth health = null;
    Set<String> customDiag;
    HashMap<String, Boolean> mapConditions;
    Boolean allowSubmission = true;
    private ChipCloudConfig config;

    private void initViews() {
        config = Utilities.getCloudConfig().selectMode(ChipCloud.SelectMode.close);
        activityAddExaminationBinding.btnAddDiag.setOnClickListener(view -> {
            customDiag.add(activityAddExaminationBinding.etOtherDiag.getText().toString());
            activityAddExaminationBinding.etOtherDiag.setText("");
            showOtherDiagnosis();
        });
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        activityAddExaminationBinding = ActivityAddExaminationBinding.inflate(getLayoutInflater());
        setContentView(activityAddExaminationBinding.getRoot());
        getSupportActionBar().setHomeButtonEnabled(true);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        customDiag = new HashSet<>();
        initViews();
        currentUser = new UserProfileDbHandler(this).getUserModel();
        mapConditions = new HashMap<String, Boolean>();
        mRealm = new DatabaseService(this).getRealmInstance();
        userId = getIntent().getStringExtra("userId");
        pojo = mRealm.where(RealmMyHealthPojo.class).equalTo("_id", userId).findFirst();
        if (pojo == null) {
            pojo = mRealm.where(RealmMyHealthPojo.class).equalTo("userId", userId).findFirst();
        }
        user = mRealm.where(RealmUserModel.class).equalTo("id", userId).findFirst();
        if (pojo != null && !TextUtils.isEmpty(pojo.getData())) {
            health = new Gson().fromJson(AndroidDecrypter.decrypt(pojo.getData(), user.getKey(), user.getIv()), RealmMyHealth.class);
        }

        if (health == null) {
            initHealth();
        }
        initExamination();
        validateFields();
        findViewById(R.id.btn_save).setOnClickListener(view -> {
            if (!isValidInput() || !allowSubmission) {
                Utilities.toast(this, getString(R.string.invalid_input));
                return;
            }
            saveData();
        });
    }

    RealmMyHealthPojo examination;

    private void initExamination() {
        if (getIntent().hasExtra("id")) {
            examination = mRealm.where(RealmMyHealthPojo.class).equalTo("_id", getIntent().getStringExtra("id")).findFirst();
            activityAddExaminationBinding.etTemperature.setText(examination.getTemperature() + "");
            activityAddExaminationBinding.etPulseRate.setText(examination.getPulse() + "");
            activityAddExaminationBinding.etBloodpressure.setText(examination.getBp() + "");
            activityAddExaminationBinding.etHeight.setText(examination.getHeight() + "");
            activityAddExaminationBinding.etWeight.setText(examination.getWeight() + "");
            activityAddExaminationBinding.etVision.setText(examination.getVision());
            activityAddExaminationBinding.etHearing.setText(examination.getHearing());
            JsonObject encrypted = examination.getEncryptedDataAsJson(this.user);
            activityAddExaminationBinding.etObservation.setText(JsonUtils.getString(getString(R.string.note_), encrypted));
            activityAddExaminationBinding.etDiag.setText(JsonUtils.getString(getString(R.string.diagno), encrypted));
            activityAddExaminationBinding.etTreatments.setText(JsonUtils.getString(getString(R.string.treat), encrypted));
            activityAddExaminationBinding.etMedications.setText(JsonUtils.getString(getString(R.string.medicay), encrypted));
            activityAddExaminationBinding.etImmunization.setText(JsonUtils.getString(getString(R.string.immunizations), encrypted));
            activityAddExaminationBinding.etAllergies.setText(JsonUtils.getString(getString(R.string.allergy), encrypted));
            activityAddExaminationBinding.etXray.setText(JsonUtils.getString(getString(R.string.xrays), encrypted));
            activityAddExaminationBinding.etLabtest.setText(JsonUtils.getString(getString(R.string.tests), encrypted));
            activityAddExaminationBinding.etReferrals.setText(JsonUtils.getString(getString(R.string.referral), encrypted));
        }
        showCheckbox(examination);
        showOtherDiagnosis();
    }

    private void validateFields() {
        allowSubmission = true;
        activityAddExaminationBinding.etBloodpressure.addTextChangedListener(new TextWatcher() {
            @Override
            public void afterTextChanged(Editable s) {
            }

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (!activityAddExaminationBinding.etBloodpressure.getText().toString().contains("/")) {
                    activityAddExaminationBinding.etBloodpressure.setError(getString(R.string.blood_pressure_should_be_numeric_systolic_diastolic));
                    allowSubmission = false;
                } else {
                    String[] sysDia = activityAddExaminationBinding.etBloodpressure.getText().toString().trim().split("/");
                    if (sysDia.length > 2 || sysDia.length < 1) {
                        activityAddExaminationBinding.etBloodpressure.setError(getString(R.string.blood_pressure_should_be_systolic_diastolic));
                        allowSubmission = false;
                    } else {
                        try {
                            int sys = Integer.parseInt(sysDia[0]);
                            int dis = Integer.parseInt(sysDia[1]);
                            if ((sys < 60 || dis < 40) || (sys > 300 || dis > 200)) {
                                activityAddExaminationBinding.etBloodpressure.setError(getString(R.string.bp_must_be_between_60_40_and_300_200));
                                allowSubmission = false;
                            } else {
                                allowSubmission = true;
                            }
                        } catch (Exception e) {
                            activityAddExaminationBinding.etBloodpressure.setError(getString(R.string.systolic_and_diastolic_must_be_numbers));
                            allowSubmission = false;
                        }
                    }
                }
            }
        });
    }

    private void showOtherDiagnosis() {
        activityAddExaminationBinding.containerOtherDiagnosis.removeAllViews();
        final ChipCloud chipCloud = new ChipCloud(this, activityAddExaminationBinding.containerOtherDiagnosis, config);

        for (String s : customDiag) {
            chipCloud.addChip(s);

        }
        chipCloud.setDeleteListener((i, s1) -> {
            customDiag.remove(s1);
        });
        preloadCustomDiagnosis(chipCloud);
    }

    private void preloadCustomDiagnosis(ChipCloud chipCloud) {
        String[] arr = getResources().getStringArray(R.array.diagnosis_list);
        List<String> mainList = Arrays.asList(arr);
        if (customDiag.isEmpty() && examination != null) {
            JsonObject conditions = new Gson().fromJson(examination.getConditions(), JsonObject.class);
            for (String s : conditions.keySet()) {
                if (!mainList.contains(s) && JsonUtils.getBoolean(s, conditions)) {
                    chipCloud.addChip(s);
                    chipCloud.setDeleteListener((i, s1) -> customDiag.remove(Constants.LABELS.get(s1)));
                    customDiag.add(s);
                }
            }
        }
    }

    private void showCheckbox(RealmMyHealthPojo examination) {
        String[] arr = getResources().getStringArray(R.array.diagnosis_list);
        activityAddExaminationBinding.containerCheckbox.removeAllViews();
        for (String s : arr) {
            CheckBox c = new CheckBox(this);
            if (examination != null) {
                JsonObject conditions = new Gson().fromJson(examination.getConditions(), JsonObject.class);
                c.setChecked(JsonUtils.getBoolean(s, conditions));
            }
            c.setPadding(DimenUtils.dpToPx(8), DimenUtils.dpToPx(8), DimenUtils.dpToPx(8), DimenUtils.dpToPx(8));
            c.setText(s);
            c.setTag(s);
            c.setOnCheckedChangeListener(this);
            activityAddExaminationBinding.containerCheckbox.addView(c);
        }
    }

    private void getOtherConditions() {
        for (String s : customDiag) {
            mapConditions.put(s, true);
        }
    }

    private void initHealth() {
        if (!mRealm.isInTransaction()) mRealm.beginTransaction();
        health = new RealmMyHealth();
        RealmMyHealth.RealmMyHealthProfile profile = new RealmMyHealth.RealmMyHealthProfile();
        health.lastExamination = new Date().getTime();
        health.userKey = AndroidDecrypter.generateKey();
        health.profile = profile;
        mRealm.commitTransaction();
    }

    private void saveData() {

        if (!mRealm.isInTransaction()) mRealm.beginTransaction();
        createPojo();
        if (examination == null) {
            String userId = AndroidDecrypter.generateIv();
            examination = mRealm.createObject(RealmMyHealthPojo.class, userId);
            examination.setUserId(userId);
        }

        examination.setProfileId(health.userKey);
        examination.setCreatorId(health.userKey);
        examination.setGender(user.getGender());

        examination.setAge(TimeUtils.getAge(user.getDob()));
        examination.setSelfExamination(currentUser.get_id().equals(pojo.get_id()));
        examination.setDate(new Date().getTime());
        examination.setPlanetCode(user.getPlanetCode());
        RealmExamination sign = new RealmExamination();
        sign.setAllergies(activityAddExaminationBinding.etAllergies.getText().toString().trim());
        sign.setCreatedBy(currentUser.get_id());
        examination.setBp(activityAddExaminationBinding.etBloodpressure.getText().toString().trim());
        examination.setTemperature(getFloat(activityAddExaminationBinding.etTemperature.getText().toString().trim()));
        examination.setPulse(getInt(activityAddExaminationBinding.etPulseRate.getText().toString().trim()));
        examination.setWeight(getFloat(activityAddExaminationBinding.etWeight.getText().toString().trim()));
        examination.setHeight(getFloat(activityAddExaminationBinding.etHeight.getText().toString().trim()));
        getOtherConditions();
        examination.setConditions(new Gson().toJson(mapConditions));
        examination.setHearing(activityAddExaminationBinding.etHearing.getText().toString().trim());
        sign.setImmunizations(activityAddExaminationBinding.etImmunization.getText().toString().trim());
        sign.setTests(activityAddExaminationBinding.etLabtest.getText().toString().trim());
        sign.setXrays(activityAddExaminationBinding.etXray.getText().toString().trim());
        examination.setVision(activityAddExaminationBinding.etVision.getText().toString().trim());
        sign.setTreatments(activityAddExaminationBinding.etTreatments.getText().toString().trim());
        sign.setReferrals(activityAddExaminationBinding.etReferrals.getText().toString().trim());
        sign.setNotes(activityAddExaminationBinding.etObservation.getText().toString().trim());
        sign.setDiagnosis(activityAddExaminationBinding.etDiag.getText().toString().trim());
        sign.setMedications(activityAddExaminationBinding.etMedications.getText().toString().trim());
        examination.setDate(new Date().getTime());
        examination.setIsUpdated(true);
        examination.setHasInfo(getHasInfo());
        pojo.setIsUpdated(true);
        try {
            Utilities.log(new Gson().toJson(sign));
            examination.setData(AndroidDecrypter.encrypt(new Gson().toJson(sign), user.getKey(), user.getIv()));
        } catch (Exception e) {
            e.printStackTrace();
        }
        mRealm.commitTransaction();
        Utilities.toast(this, getString(R.string.added_successfully));
        super.finish();
    }

    private boolean getHasInfo() {
        return !TextUtils.isEmpty(activityAddExaminationBinding.etAllergies.getText().toString()) ||
                !TextUtils.isEmpty(activityAddExaminationBinding.etDiag.getText().toString()) ||
                !TextUtils.isEmpty(activityAddExaminationBinding.etImmunization.getText().toString()) ||
                !TextUtils.isEmpty(activityAddExaminationBinding.etMedications.getText().toString()) ||
                !TextUtils.isEmpty(activityAddExaminationBinding.etObservation.getText().toString()) ||
                !TextUtils.isEmpty(activityAddExaminationBinding.etReferrals.getText().toString()) ||
                !TextUtils.isEmpty(activityAddExaminationBinding.etLabtest.getText().toString()) ||
                !TextUtils.isEmpty(activityAddExaminationBinding.etTreatments.getText().toString()) ||
                !TextUtils.isEmpty(activityAddExaminationBinding.etXray.getText().toString());
    }

    private boolean isValidInput() {
        boolean isValidTemp = 30 <= getFloat(activityAddExaminationBinding.etTemperature.getText().toString().trim()) && getFloat(activityAddExaminationBinding.etTemperature.getText().toString().trim()) <= 40 || getFloat(activityAddExaminationBinding.etTemperature.getText().toString().trim()) == 0;
        boolean isValidPulse = 40 <= getInt(activityAddExaminationBinding.etPulseRate.getText().toString().trim()) && getInt(activityAddExaminationBinding.etPulseRate.getText().toString().trim()) <= 120 || getFloat(activityAddExaminationBinding.etPulseRate.getText().toString().trim()) == 0;
        boolean isValidHeight = 1 <= getFloat(activityAddExaminationBinding.etHeight.getText().toString().trim()) && getFloat(activityAddExaminationBinding.etHeight.getText().toString().trim()) <= 250 || getFloat(activityAddExaminationBinding.etHeight.getText().toString().trim()) == 0;
        boolean isValidWeight = 1 <= getFloat(activityAddExaminationBinding.etWeight.getText().toString().trim()) && getFloat(activityAddExaminationBinding.etWeight.getText().toString().trim()) <= 150 || getFloat(activityAddExaminationBinding.etWeight.getText().toString().trim()) == 0;
        if (!isValidTemp) {
            activityAddExaminationBinding.etTemperature.setError(getString(R.string.invalid_input_must_be_between_30_and_40));
        }
        if (!isValidPulse) {
            activityAddExaminationBinding.etPulseRate.setError(getString(R.string.invalid_input_must_be_between_40_and_120));
        }
        if (!isValidHeight) {
            activityAddExaminationBinding.etHeight.setError(getString(R.string.invalid_input_must_be_between_1_and_250));
        }
        if (!isValidWeight) {
            activityAddExaminationBinding.etWeight.setError(getString(R.string.invalid_input_must_be_between_1_and_150));
        }
        return isValidTemp && isValidHeight && isValidPulse && isValidWeight;
    }

//    private float getFloat(String trim) {
//    }

    private int getInt(String trim) {
        try {
            return Integer.parseInt(trim);
        } catch (Exception e) {
            return 0;
        }
    }

    private float getFloat(String trim) {
        try {
            return Float.parseFloat(String.format("%.1f", Float.parseFloat(trim)));
        } catch (Exception e) {
            return getInt(trim);
        }
    }

    private void createPojo() {
        try {
            if (pojo == null) {
                pojo = mRealm.createObject(RealmMyHealthPojo.class, userId);
                pojo.setUserId(user.get_id());
            }
//            if (TextUtils.isEmpty(pojo.getData())) {
            health.lastExamination = new Date().getTime();
            pojo.setData(AndroidDecrypter.encrypt(new Gson().toJson(health), user.getKey(), user.getIv()));
//            }
        } catch (Exception e) {
            e.printStackTrace();
            Utilities.toast(this, getString(R.string.unable_to_add_health_record));
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void finish() {
        AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(this);
        alertDialogBuilder.setMessage(R.string.are_you_sure_you_want_to_exit_your_data_will_be_lost);
        alertDialogBuilder.setPositiveButton(getString(R.string.yes_i_want_to_exit), (dialogInterface, i) -> {
            super.finish();
        }).setNegativeButton(getString(R.string.cancel), null);
        alertDialogBuilder.show();
    }

    @Override
    public void onCheckedChanged(CompoundButton compoundButton, boolean b) {
        String text = compoundButton.getText().toString().trim();
        mapConditions.put(text, b);
    }
}

File: app/src/main/java/org/ole/planet/myplanet/ui/myhealth/AddExaminationActivity.kt
package org.ole.planet.myplanet.ui.myhealth

import android.content.DialogInterface
import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.view.MenuItem
import android.view.View
import android.widget.CheckBox
import android.widget.CompoundButton
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.google.gson.JsonObject
import dagger.hilt.android.AndroidEntryPoint
import fisk.chipcloud.ChipCloud
import fisk.chipcloud.ChipCloudConfig
import io.realm.Realm
import java.util.Date
import java.util.Locale
import javax.inject.Inject
import org.ole.planet.myplanet.R
import org.ole.planet.myplanet.databinding.ActivityAddExaminationBinding
import org.ole.planet.myplanet.model.RealmExamination
import org.ole.planet.myplanet.model.RealmMyHealth
import org.ole.planet.myplanet.model.RealmMyHealthPojo
import org.ole.planet.myplanet.model.RealmUserModel
import org.ole.planet.myplanet.service.UserProfileDbHandler
import org.ole/planet.myplanet.utilities.AndroidDecrypter.Companion.decrypt
import org.ole.planet.myplanet.utilities.AndroidDecrypter.Companion.encrypt
import org.ole.planet.myplanet.utilities.AndroidDecrypter.Companion.generateIv
import org.ole.planet.myplanet.utilities.AndroidDecrypter.Companion.generateKey
import org.ole.planet.myplanet.utilities.Constants
import org.ole.planet.myplanet.utilities.DimenUtils.dpToPx
import org.ole.planet.myplanet.utilities.EdgeToEdgeUtils
import org.ole.planet.myplanet.utilities.GsonUtils
import org.ole.planet.myplanet.utilities.JsonUtils.getBoolean
import org.ole.planet.myplanet.utilities.JsonUtils.getString
import org.ole.planet.myplanet.utilities.TimeUtils.getAge
import org.ole.planet.myplanet.utilities.Utilities

import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.launch
import org.ole.planet.myplanet.repository.ExaminationRepository

@AndroidEntryPoint
class AddExaminationActivity : AppCompatActivity(), CompoundButton.OnCheckedChangeListener {
    @Inject
    lateinit var examinationRepository: ExaminationRepository
    @Inject
    lateinit var userProfileDbHandler: UserProfileDbHandler
    private lateinit var binding: ActivityAddExaminationBinding
    var userId: String? = null
    var user: RealmUserModel? = null
    private var currentUser: RealmUserModel? = null
    private var pojo: RealmMyHealthPojo? = null
    var health: RealmMyHealth? = null
    private var customDiag: MutableSet<String?>? = null
    private var mapConditions: HashMap<String?, Boolean>? = null
    var allowSubmission = true
    private lateinit var config: ChipCloudConfig
    private var examination: RealmMyHealthPojo? = null
    private var textWatcher: TextWatcher? = null
    private fun initViews() {
        config = Utilities.getCloudConfig().selectMode(ChipCloud.SelectMode.close)
        binding.btnAddDiag.setOnClickListener {
            customDiag?.add("${binding.etOtherDiag.text}")
            binding.etOtherDiag.setText(R.string.empty_text)
            showOtherDiagnosis()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityAddExaminationBinding.inflate(layoutInflater)
        setContentView(binding.root)
        EdgeToEdgeUtils.setupEdgeToEdgeWithKeyboard(this, binding.root)
        supportActionBar?.setHomeButtonEnabled(true)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        customDiag = HashSet()
        initViews()
        currentUser = userProfileDbHandler.userModel
        mapConditions = HashMap()
        userId = intent.getStringExtra("userId")

        lifecycleScope.launch {
            val (loadedPojo, loadedUser) = examinationRepository.getHealthAndUser(userId ?: "")
            pojo = loadedPojo
            user = loadedUser
            health = examinationRepository.getDecryptedHealth(pojo, user)
            initExamination()
        }

        validateFields()
        findViewById<View>(R.id.btn_save).setOnClickListener {
            if (!allowSubmission) {
                scrollToView(binding.etBloodpressure)
            }
            if (isValidInput && allowSubmission) {
                saveData()
            }
        }
    }

    private fun initExamination() {
        if (intent.hasExtra("id")) {
            lifecycleScope.launch {
                examination = examinationRepository.getExamination(intent.getStringExtra("id")!!)
                binding.etTemperature.setText(getString(R.string.float_placeholder, examination?.temperature))
                binding.etPulseRate.setText(getString(R.string.number_placeholder, examination?.pulse))
                binding.etBloodpressure.setText(getString(R.string.message_placeholder, examination?.bp))
            binding.etHeight.setText(getString(R.string.float_placeholder, examination?.height))
            binding.etWeight.setText(getString(R.string.float_placeholder, examination?.weight))
            binding.etVision.setText(examination?.vision)
            binding.etHearing.setText(examination?.hearing)
            val encrypted = user?.let { examination?.getEncryptedDataAsJson(it) }
            binding.etObservation.setText(getString(getString(R.string.note_), encrypted))
            binding.etDiag.setText(getString(getString(R.string.diagnosis), encrypted))
            binding.etTreatments.setText(getString(getString(R.string.treatments), encrypted))
            binding.etMedications.setText(getString(getString(R.string.medications), encrypted))
            binding.etImmunization.setText(getString(getString(R.string.immunizations), encrypted))
            binding.etAllergies.setText(getString(getString(R.string.allergies), encrypted))
            binding.etXray.setText(getString(getString(R.string.xrays), encrypted))
            binding.etLabtest.setText(getString(getString(R.string.tests), encrypted))
            binding.etReferrals.setText(getString(getString(R.string.referrals), encrypted))
            }
        }
        showCheckbox(examination)
        showOtherDiagnosis()
    }

    private fun validateFields() {
        allowSubmission = true
        textWatcher = object : TextWatcher {
            override fun afterTextChanged(s: Editable) {}
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                if (!"${binding.etBloodpressure.text}".contains("/")) {
                    binding.etBloodpressure.error = getString(R.string.blood_pressure_should_be_numeric_systolic_diastolic)
                    allowSubmission = false
                } else {
                    val sysDia = "${binding.etBloodpressure.text}"
                        .trim { it <= ' ' }
                        .split("/".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                    if (sysDia.size > 2 || sysDia.isEmpty()) {
                        binding.etBloodpressure.error = getString(R.string.blood_pressure_should_be_systolic_diastolic)
                        allowSubmission = false
                    } else {
                        try {
                            val sys = sysDia[0].toInt()
                            val dis = sysDia[1].toInt()
                            if (sys < 60 || dis < 40 || sys > 300 || dis > 200) {
                                binding.etBloodpressure.error = getString(R.string.bp_must_be_between_60_40_and_300_200)
                                allowSubmission = false
                            } else {
                                allowSubmission = true
                            }
                        } catch (e: Exception) {
                            binding.etBloodpressure.error = getString(R.string.systolic_and_diastolic_must_be_numbers)
                            allowSubmission = false
                        }
                    }
                }
            }
        }
        binding.etBloodpressure.addTextChangedListener(textWatcher)
    }

    private fun showOtherDiagnosis() {
        binding.containerOtherDiagnosis.removeAllViews()
        val chipCloud = ChipCloud(this, binding.containerOtherDiagnosis, config)
        for (s in customDiag?: emptySet()) {
            if (s.isNullOrBlank()) {
                    continue
            } else {
                    chipCloud.addChip(s)
            }
        }
        chipCloud.setDeleteListener { _: Int, s1: String? -> customDiag?.remove(s1) }
        preloadCustomDiagnosis(chipCloud)
    }

    private fun preloadCustomDiagnosis(chipCloud: ChipCloud) {
        val arr = resources.getStringArray(R.array.diagnosis_list)
        val mainList = listOf(*arr)
        if (customDiag?.isEmpty() == true && examination != null) {
            val conditions = GsonUtils.gson.fromJson(examination?.conditions, JsonObject::class.java)
            for (s in conditions.keySet()) {
                if (!mainList.contains(s) && getBoolean(s, conditions)) {
                    chipCloud.addChip(s)
                    chipCloud.setDeleteListener { _: Int, s1: String? ->
                        customDiag?.remove(Constants.LABELS[s1])
                    }
                    customDiag?.add(s)
                }
            }
        }
    }

    private fun showCheckbox(examination: RealmMyHealthPojo?) {
        val arr = resources.getStringArray(R.array.diagnosis_list)
        binding.containerCheckbox.removeAllViews()
        for (s in arr) {
            val c = CheckBox(this)
            c.buttonTintList = ContextCompat.getColorStateList(this, R.color.daynight_textColor)
            c.setTextColor(ContextCompat.getColor(this, R.color.daynight_textColor))

            if (examination != null) {
                val conditions = GsonUtils.gson.fromJson(examination.conditions, JsonObject::class.java)
                c.isChecked = getBoolean(s, conditions)
            }
            c.setPadding(dpToPx(8), dpToPx(8), dpToPx(8), dpToPx(8))
            c.text = s
            c.tag = s
            c.setOnCheckedChangeListener(this)
            binding.containerCheckbox.addView(c)
        }
    }

    private val otherConditions: Unit
        get() {
            for (s in customDiag?: emptySet()) {
                mapConditions?.set(s, true)
            }
        }

    private fun saveData() {
        val sign = RealmExamination().apply {
            allergies = "${binding.etAllergies.text}".trim()
            createdBy = currentUser?._id
            immunizations = "${binding.etImmunization.text}".trim()
            tests = "${binding.etLabtest.text}".trim()
            xrays = "${binding.etXray.text}".trim()
            treatments = "${binding.etTreatments.text}".trim()
            referrals = "${binding.etReferrals.text}".trim()
            notes = "${binding.etObservation.text}".trim()
            diagnosis = "${binding.etDiag.text}".trim()
            medications = "${binding.etMedications.text}".trim()
        }
        otherConditions

        lifecycleScope.launch {
            val success = examinationRepository.saveExamination(
                examination,
                health,
                user,
                currentUser,
                sign,
                mapConditions ?: emptyMap(),
                getFloat("${binding.etTemperature.text}".trim()),
                getInt("${binding.etPulseRate.text}".trim()),
                getFloat("${binding.etHeight.text}".trim()),
                getFloat("${binding.etWeight.text}".trim()),
                "${binding.etVision.text}".trim(),
                "${binding.etHearing.text}".trim(),
                "${binding.etBloodpressure.text}".trim(),
                hasInfo
            )
            if (success) {
                Utilities.toast(this@AddExaminationActivity, getString(R.string.added_successfully))
                super.finish()
            } else {
                Utilities.toast(this@AddExaminationActivity, getString(R.string.unable_to_add_health_record))
            }
        }
    }

    private fun scrollToView(view: View) {
        binding.rootScrollView.post {
            binding.rootScrollView.smoothScrollTo(0, view.top)
            view.requestFocus()
        }
    }

    private val hasInfo: Boolean
        get() = !TextUtils.isEmpty("${binding.etAllergies.text}") ||
                !TextUtils.isEmpty("${binding.etDiag.text}") ||
                !TextUtils.isEmpty("${binding.etImmunization.text}") ||
                !TextUtils.isEmpty("${binding.etMedications.text}") ||
                !TextUtils.isEmpty("${binding.etObservation.text}") ||
                !TextUtils.isEmpty("${binding.etReferrals.text}") ||
                !TextUtils.isEmpty("${binding.etLabtest.text}") ||
                !TextUtils.isEmpty("${binding.etTreatments.text}") ||
                !TextUtils.isEmpty("${binding.etXray.text}")
    private val isValidInput: Boolean
        get() {
            val scrollView = binding.rootScrollView

            val isValidTemp = (getFloat("${binding.etTemperature.text}".trim { it <= ' ' }) in 30.0..40.0 ||
                        getFloat("${binding.etTemperature.text}".trim { it <= ' ' }) == 0f) &&
                    "${binding.etTemperature.text}".trim { it <= ' ' }.isNotEmpty()
            val isValidPulse = (getInt("${binding.etPulseRate.text}".trim { it <= ' ' }) in 40..120 ||
                    getFloat("${binding.etPulseRate.text}".trim { it <= ' ' }) == 0f) &&
                    "${binding.etPulseRate.text}".trim { it <= ' ' }.isNotEmpty()
            val isValidHeight = (getFloat("${binding.etHeight.text}".trim { it <= ' ' }) in 1.0..250.0 ||
                    getFloat("${binding.etHeight.text}".trim { it <= ' ' }) == 0f) &&
                    "${binding.etHeight.text}".trim { it <= ' ' }.isNotEmpty()
            val isValidWeight = (getFloat("${binding.etWeight.text}".trim { it <= ' ' }) in 1.0..150.0 ||
                    getFloat("${binding.etWeight.text}".trim { it <= ' ' }) == 0f) &&
                    "${binding.etWeight.text}".trim { it <= ' ' }.isNotEmpty()
            if (!isValidTemp) {
                binding.etTemperature.error = getString(R.string.invalid_input_must_be_between_30_and_40)
                scrollToView(binding.etTemperature)
                Utilities.toast(this, getString(R.string.invalid_input_must_be_between_30_and_40))
            }
            if (!isValidPulse) {
                binding.etPulseRate.error = getString(R.string.invalid_input_must_be_between_40_and_120)
                Utilities.toast(this, getString(R.string.invalid_input_must_be_between_40_and_120))
                scrollToView(binding.etPulseRate)
            }
            if (!isValidHeight) {
                binding.etHeight.error = getString(R.string.invalid_input_must_be_between_1_and_250)
                Utilities.toast(this, getString(R.string.invalid_input_must_be_between_1_and_250))
                scrollToView(binding.etHeight)
            }
            if (!isValidWeight) {
                binding.etWeight.error = getString(R.string.invalid_input_must_be_between_1_and_150)
                Utilities.toast(this, getString(R.string.invalid_input_must_be_between_1_and_150))
                scrollToView(binding.etWeight)
            }
            return isValidTemp && isValidHeight && isValidPulse && isValidWeight
        }

    //    private float getFloat(String trim) {
    //    }
    private fun getInt(trim: String): Int {
        return try {
            trim.toInt()
        } catch (e: Exception) {
            0
        }
    }

    private fun getFloat(trim: String): Float {
        return try {
            String.format(Locale.getDefault(), "%.1f", trim.toFloat()).toFloat()
        } catch (e: Exception) {
            getInt(trim).toFloat()
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun finish() {
        val alertDialogBuilder = AlertDialog.Builder(this, R.style.AlertDialogTheme)
        alertDialogBuilder.setMessage(R.string.cancel_adding_examination)
        alertDialogBuilder.setPositiveButton(getString(R.string.yes_i_want_to_exit)) { _: DialogInterface?, _: Int -> super.finish() }
            .setNegativeButton(getString(R.string.cancel), null)
        alertDialogBuilder.show()
    }

    override fun onCheckedChanged(compoundButton: CompoundButton, b: Boolean) {
        val text = "${compoundButton.text}".trim { it <= ' ' }
        mapConditions?.set(text, b)
    }

    override fun onDestroy() {
        binding.etBloodpressure.removeTextChangedListener(textWatcher)
        textWatcher = null
        super.onDestroy()
    }
}
